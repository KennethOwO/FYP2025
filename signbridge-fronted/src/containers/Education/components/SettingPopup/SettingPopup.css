.setting-container {
    position: fixed;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    background-color: rgba(0, 0, 0, 0.5);
    display: flex;
    justify-content: center;
    align-items: center;
    z-index: 999;
}

.setting-content {
    background-color: #4c4e4d;
    border-radius: 20px;
    box-shadow: 0 0 10px rgba(0, 0, 0, 0.3);
    overflow: auto;
}

.setting-content h3 {
    font-family: "Russo One", sans-serif;
    background-color: #4c4e4d;
    padding: 20px 20px 10px 20px;
    color: #f4f4f4;
    text-align: center;
    font-size: 1.9em;
}

.setting-details {
    background-color: #ecedec;
    border: 8px solid #4c4e4d;
    border-radius: 20px;
    width: 500px;
    height: 160px;
    padding: 20px;
}

.setting-details p {
    margin: 20px 0;
}

.setting-content p {
    font-family: "Russo One", sans-serif;
    padding: 0 40px;
    font-size: 20px;
}

/* Volumn Slider */
.volume-title {
    font-family: "Russo One", sans-serif;
    font-size: 20px;
}

.slider-value-label {
    color: black;
    display: inline-block !important;
    font-size: 22px !important;
    margin-top: 7px !important;
    text-align: center;
    width: 100%;
}

.noUi-handle {
    border: none !important;
    border-radius: 0 !important;
    background: url("https://raw.githubusercontent.com/cromartie1984/diablo3-project/master/ScrollButton.png") !important;
    background-position: left top !important;
    box-shadow: none !important;
}

.noUi-handle:focus {
    outline: none !important;
}

.noUi-handle:hover {
    background-position: left center !important;
}

.noUi-handle:active {
    background-position: left bottom !important;
}

.noUi-horizontal .noUi-handle {
    width: 25px !important;
    height: 35px !important;
    top: -12px !important;
    cursor: pointer;
}

.noUi-handle::after,
.noUi-handle::before {
    background: none !important;
}

.noUi-connect {
    box-shadow: none !important;
    border-radius: 0 !important;
    background: #f1793c !important;
    background: -moz-radial-gradient(center, ellipse cover, #f1793c 0%, #cb5e26 50%, #792f14 100%) !important;
    background: -webkit-radial-gradient(center, ellipse cover, #f1793c 0%, #cb5e26 50%, #792f14 100%) !important;
    background: radial-gradient(ellipse at center, #f1793c 0%, #cb5e26 50%, #792f14 100%) !important;
    filter: progid:DXImageTransform.Microsoft.gradient(
		startColorstr="#f1793c",
		endColorstr="#792f14",
		GradientType=1
    ) !important;
}

.noUi-target {
    box-shadow: none !important;
    width: 85% !important;
    height: 20px !important;
    border: none !important;
    border-radius: 0 !important;
    background: none !important;
    border-style: solid !important;
    border-width: 5px 4px 5px 4px !important;
    -moz-border-image: url("https://raw.githubusercontent.com/cromartie1984/diablo3-project/master/SliderBar.png") 5 4 5 4 fill stretch repeat !important;
    -webkit-border-image: url("https://raw.githubusercontent.com/cromartie1984/diablo3-project/master/SliderBar.png") 5 4 5 4 fill stretch repeat !important;
    -o-border-image: url("https://raw.githubusercontent.com/cromartie1984/diablo3-project/master/SliderBar.png") 5 4 5 4 fill stretch repeat !important;
    border-image: url("https://raw.githubusercontent.com/cromartie1984/diablo3-project/master/SliderBar.png") 5 4 5 4 fill stretch repeat !important;
}

.slider-layout {
    display: flex;
    justify-content: center;
    align-items: center;
    margin-top: 20px;
}

.fa-close-btn {
    font-size: 22px;
}

.fa-volume-down {
    padding-right: 10px;
}

.volume-icon-setting {
    padding-left: 20px;
}

@media (max-width: 768px) {
    .setting-details {
        width: 100%;
        height: auto;
    }

    .setting-content {
        max-width: 90%;
        width: 90%;
    }

    .setting-content p {
        padding: 0 20px;
        font-size: 1em;
    }
}

@media (max-width: 480px) {
    .setting-details {
        width: 100%;
    }

    .setting-content {
        max-width: 90%;
        width: 90%;
    }

    .setting-content p {
        padding: 0 10px;
        font-size: 0.9em;
    }
}
