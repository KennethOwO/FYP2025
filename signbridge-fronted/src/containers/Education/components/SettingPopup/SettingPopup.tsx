import React, { useEffect, useRef, useState } from "react";
import buttonClickedSound from "/music/btnClicked.wav";
import noUiSlider from "nouislider";
import wNumb from "wnumb";
import "nouislider/dist/nouislider.css";
import "./SettingPopup.css";
import { useTranslation } from "react-i18next";

interface SettingPopupProps {
	onClose: () => void;
	onVolumeChange: (volume: number) => void;
}

const SettingPopup: React.FC<SettingPopupProps> = ({ onClose, onVolumeChange }) => {
	const { t, i18n } = useTranslation();
	const popupRef = useRef<HTMLDivElement>(null);
	const [volume, setVolume] = useState(() => {
		const storedVolume = localStorage.getItem("volumeValue");
		return storedVolume ? parseInt(storedVolume, 10) : 100;
	});

	const playButtonClickedSound = () => {
		const audio = new Audio(buttonClickedSound);
		audio.play();
	};

	const ambientSliderRef = useRef<HTMLDivElement | null>(null);
	// @ts-ignore
	const ambientSlider = useRef<noUiSlider.API | null>(null);

	useEffect(() => {
		const ambientSliderElement = ambientSliderRef.current;
		if (ambientSliderElement) {
			const formatOptions = { decimals: 0 };
			if (ambientSlider.current !== null) {
				ambientSlider.current.destroy();
			}
			ambientSlider.current = noUiSlider.create(ambientSliderElement, {
				start: [volume],
				behaviour: "snap",
				animate: true,
				animationDuration: 300,
				orientation: "horizontal",
				connect: [true, false],
				step: 5,
				range: {
					min: [0],
					max: [100],
				},
				format: wNumb(formatOptions),
			});
			ambientSlider.current.on("slide", (values: string[]) => {
				const newVolume = parseInt(values[0]);
				setVolume(newVolume);
				localStorage.setItem("volumeValue", newVolume.toString());
				onVolumeChange(newVolume / 100);
			});
		}

		const handleClickOutside = (event: MouseEvent) => {
			if (popupRef.current && !popupRef.current.contains(event.target as Node)) {
				onClose();
			}
		};

		document.addEventListener("mousedown", handleClickOutside);

		return () => {
			document.removeEventListener("mousedown", handleClickOutside);
		};
	}, [onClose, onVolumeChange, volume]);

	return (
		<div className="setting-container">
			<div ref={popupRef} className="setting-content">
				<h3>{t('setting')}</h3>
				<button
					className="close_btn_pushable"
					role="button"
					onClick={() => {
						onClose();
						playButtonClickedSound();
					}}>
					<span className="close_btn_shadow"></span>
					<span className="close_btn_edge"></span>
					<span className="close_btn_front">
						<i className="fa fa-close fa-close-btn"></i>
					</span>
				</button>
				<div className="setting-details">
					<h4 className="volume-title">{t('music_volume')}</h4>
					<div>
						<div className="slider-layout">
							<i className="fa fa-volume-down"></i>
							<div ref={ambientSliderRef} id="ambient-volume" className="volume-slider" />
							<div className="volume-icon-setting">
								<i className="fa fa-volume-up"></i>
							</div>
						</div>
						<div className="slider-value-label" id="ambient-volume-label">
							{volume}
						</div>
					</div>
				</div>
			</div>
		</div>
	);
};

export default SettingPopup;
