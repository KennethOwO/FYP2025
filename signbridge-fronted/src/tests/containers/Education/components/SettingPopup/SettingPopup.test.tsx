import SettingPopup from "@root/containers/Education/components/SettingPopup/SettingPopup";
import { create } from "react-test-renderer";
import { cleanup } from "@testing-library/react";
jest.mock("react-i18next", () => ({
  useTranslation: () => ({
    t: (key: string) => key, // Mock the translation function to return the key
    i18n: {
      changeLanguage: jest.fn(), // Mock the changeLanguage function
    },
  }),
}));
describe("Test SettingPopup", () => {
  afterEach(() => {
    // Clean up the DOM
    cleanup();
    // Reset mocks to their initial state
    jest.resetAllMocks();
  });
  it("should render correctly", () => {
    const props = {
      onClose: jest.fn(),
      onVolumeChange: jest.fn(),
    };
    const tree = create(<SettingPopup {...props} />);
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
