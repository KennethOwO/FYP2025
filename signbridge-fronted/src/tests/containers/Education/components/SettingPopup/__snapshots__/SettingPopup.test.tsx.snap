// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test SettingPopup should render correctly 1`] = `
<div
  className="setting-container"
>
  <div
    className="setting-content"
  >
    <h3>
      setting
    </h3>
    <button
      className="close_btn_pushable"
      onClick={[Function]}
      role="button"
    >
      <span
        className="close_btn-shadow"
      />
      <span
        className="close_btn_edge"
      />
      <span
        className="close_btn_front text"
      >
        <i
          className="fa fa-close fa-close-btn"
        />
      </span>
    </button>
    <div
      className="setting-details"
    >
      <h4
        className="volume-title"
      >
        music_volume
      </h4>
      <div
        className="game-volume-bar-container"
      >
        <div
          className="slider-layout"
        >
          <i
            className="fa fa-volume-down"
          />
          <div
            className="volume-slider"
            id="ambient-volume"
          />
          <i
            className="fa fa-volume-up"
          />
        </div>
        <div
          className="slider-value-label float-left"
          id="ambient-volume-label"
        >
          100
        </div>
      </div>
    </div>
  </div>
</div>
`;
